### PR TITLE
Add new `yolov8-ghost.yaml` model

### PR DESCRIPTION
**Pull request by Burhan-Q**
# Summary

Adds configuration YAML. Config file shared by @wbwb014 on [Discord](https://ultralytics.com/discord). Using `YOLOv8-ghost` may be a good option for more lightweight deployments on the edge where the smaller number of parameters will help reduce the model weights file size and reduce inference times (or require less compute).

## Testing & Performance

All configurations run using [Ultralytics VisDrone dataset](https://docs.ultralytics.com/datasets/detect/visdrone/). For 100 epochs, with `batch=8`, using default image size `640`. Final validation results using 584 images in `val` and 1610 images for test dataset from `test`. Test dataset images used the same arguments (`batch=8`, `device='cuda'`).

### YOLOv8n-ghost from scratch

```py
from ultralytics import YOLO

model = YOLO('yolov8n-ghost.yaml')
_ = model.to('cuda')
model.reset_weights()
results = model.train(data='VisDrone.yaml', epochs=100, batch=8)
```

<details>
<summary>Final validation results</summary>

```
100 epochs completed in 3.050 hours.
Optimizer stripped from runs/detect/train14/weights/last.pt, 3.7MB
Optimizer stripped from runs/detect/train14/weights/best.pt, 3.7MB

Validating runs/detect/train14/weights/best.pt...
Ultralytics YOLOv8.0.215 🚀 Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 2060, 5924MiB)
YOLOv8n-ghost summary (fused): 312 layers, 1716026 parameters, 0 gradients, 5.0 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)
                   all        548      38759      0.368      0.278      0.263      0.146
            pedestrian        548       8844      0.343      0.315      0.276      0.111
                people        548       5125      0.458      0.198      0.224     0.0753
               bicycle        548       1287      0.184     0.0381     0.0432     0.0152
                   car        548      14064      0.549      0.727      0.709      0.468
                   van        548       1975      0.399      0.313      0.297      0.196
                 truck        548        750      0.361      0.205      0.196       0.12
              tricycle        548       1045       0.32      0.183       0.15     0.0786
       awning-tricycle        548        532      0.229      0.118     0.0924     0.0562
                   bus        548        251      0.452      0.355      0.361      0.223
                 motor        548       4886      0.385      0.326      0.286      0.113

Speed: 0.1ms preprocess, 1.3ms inference, 0.0ms loss, 1.8ms postprocess per image
```

</details>

<details>
<summary>Test dataset performance</summary>

```
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):
                   all       1610      75102      0.307       0.24      0.211      0.113
            pedestrian       1610      21006      0.338      0.187      0.173      0.065
                people       1610       6376      0.329     0.0729     0.0908     0.0296
               bicycle       1610       1302      0.173     0.0315     0.0386     0.0145
                   car       1610      28074      0.497      0.645      0.602      0.353
                   van       1610       5771      0.263      0.312      0.241      0.146
                 truck       1610       2659      0.277      0.269      0.203      0.113
              tricycle       1610        530      0.134      0.111      0.071     0.0342
       awning-tricycle       1610        599      0.278       0.11      0.105     0.0539
                   bus       1610       2940      0.468      0.435      0.413      0.255
                 motor       1610       5845      0.315      0.227      0.173     0.0638

Speed: 0.1ms preprocess, 2.1ms inference, 0.0ms loss, 1.1ms postprocess per image
```

</details>

### YOLOv8n COCO pretrained

```py
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
_ = model.to('cuda')
results = model.train(data='VisDrone.yaml', epochs=100, batch=8)
```

<details>
<summary>Final validation results</summary>

```
100 epochs completed in 2.659 hours.
Optimizer stripped from runs/detect/train15/weights/last.pt, 6.2MB
Optimizer stripped from runs/detect/train15/weights/best.pt, 6.2MB

Validating runs/detect/train15/weights/best.pt...
Ultralytics YOLOv8.0.215  Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 2060, 5924MiB)
Model summary (fused): 168 layers, 3007598 parameters, 0 gradients, 8.1 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):
                   all        548      38759      0.441      0.331      0.335      0.195
            pedestrian        548       8844       0.44      0.341       0.35       0.15
                people        548       5125      0.508      0.227      0.283      0.103
               bicycle        548       1287      0.239      0.108     0.0826      0.032
                   car        548      14064      0.645      0.747      0.757       0.52
                   van        548       1975      0.493      0.385      0.392      0.273
                 truck        548        750       0.44      0.326      0.312      0.209
              tricycle        548       1045      0.403      0.259      0.239      0.131
       awning-tricycle        548        532      0.241      0.148      0.116     0.0748
                   bus        548        251      0.534       0.42      0.456       0.31
                 motor        548       4886      0.463      0.352      0.359       0.15

Speed: 0.1ms preprocess, 1.3ms inference, 0.0ms loss, 1.2ms postprocess per image
```

</details>

<details>
<summary>Test dataset performance</summary>

```
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):
                   all       1610      75102      0.402      0.293      0.277      0.156
            pedestrian       1610      21006      0.451      0.216      0.225     0.0872
                people       1610       6376      0.459     0.0877       0.12     0.0412
               bicycle       1610       1302      0.229     0.0699     0.0621      0.025
                   car       1610      28074      0.618       0.68       0.67      0.408
                   van       1610       5771      0.358      0.349      0.308      0.194
                 truck       1610       2659      0.365      0.398      0.347      0.217
              tricycle       1610        530      0.224      0.213      0.147     0.0791
       awning-tricycle       1610        599      0.333      0.157      0.152     0.0807
                   bus       1610       2940      0.603      0.501      0.517      0.346
                 motor       1610       5845      0.382      0.259      0.221     0.0824

Speed: 0.1ms preprocess, 2.9ms inference, 0.0ms loss, 1.1ms postprocess per image
```

</details>

### YOLOv8n from scratch

```py
from ultralytics import YOLO

model = YOLO('yolov8n.yaml')
_ = model.to('cuda')
model.reset_weights()
results = model.train(data='VisDrone.yaml', epochs=100, batch=8)
```

<details>
<summary>Final validation results</summary>

```
100 epochs completed in 2.665 hours.
Optimizer stripped from runs/detect/train16/weights/last.pt, 6.2MB
Optimizer stripped from runs/detect/train16/weights/best.pt, 6.2MB

Validating runs/detect/train16/weights/best.pt...
Ultralytics YOLOv8.0.215  Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 2060, 5924MiB)
YOLOv8n summary (fused): 168 layers, 3007598 parameters, 0 gradients, 8.1 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):
                   all        548      38759      0.409      0.312      0.305      0.174
            pedestrian        548       8844      0.406      0.325      0.313      0.129
                people        548       5125      0.485      0.208       0.25     0.0869
               bicycle        548       1287      0.198     0.0808     0.0695      0.025
                   car        548      14064      0.636      0.737       0.74      0.503
                   van        548       1975      0.441      0.371      0.367      0.247
                 truck        548        750      0.394      0.261      0.246      0.158
              tricycle        548       1045       0.36      0.236      0.206      0.111
       awning-tricycle        548        532      0.227      0.135      0.106     0.0666
                   bus        548        251      0.518      0.406      0.416      0.275
                 motor        548       4886      0.427      0.356      0.335      0.134

Speed: 0.1ms preprocess, 1.3ms inference, 0.0ms loss, 1.5ms postprocess per image
```

</details>

<details>
<summary>Test dataset performance</summary>

```
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):
                   all       1610      75102      0.348      0.277      0.247      0.137
            pedestrian       1610      21006      0.407      0.199      0.202     0.0777
                people       1610       6376      0.392     0.0743      0.103     0.0333
               bicycle       1610       1302      0.184     0.0499      0.049     0.0204
                   car       1610      28074      0.567      0.662      0.638      0.386
                   van       1610       5771      0.284      0.367      0.285      0.177
                 truck       1610       2659       0.32      0.349      0.286       0.17
              tricycle       1610        530      0.189      0.189      0.105     0.0537
       awning-tricycle       1610        599      0.299      0.139      0.122     0.0637
                   bus       1610       2940      0.512      0.488      0.477      0.309
                 motor       1610       5845      0.328      0.251      0.201     0.0751

Speed: 0.2ms preprocess, 3.0ms inference, 0.0ms loss, 2.0ms postprocess per image
```

</details>

### Final validation (all classes)

| Model structure | Pretrained | Box( P ) | Box( R ) | Box( mAP50 ) | Box( mAP50-95) ) |
| :----------------- | :----------: | :--------: |:--------: | :--------------: | :------------------: |
| YOLOv8n-ghost |        n        |    0.368   |   0.278  |        0.263        |          0.146           |
| YOLOv8n            |        y        |    0.441   |   0.331  |        0.335        |          0.195           |
| YOLOv8n            |        n        |    0.409   |   0.312  |        0.305        |          0.174           |

### Test dataset

| Model structure | Pretrained | Box( P ) | Box( R ) | Box( mAP50 ) | Box( mAP50-95) ) |
| :----------------- | :----------: | :--------: |:--------: | :--------------: | :------------------: |
| YOLOv8n-ghost |        n        |    0.307   |   0.240  |        0.211        |          0.113           |
| YOLOv8n            |        y        |    0.402   |   0.293  |        0.277        |          0.156           |
| YOLOv8n            |        n        |    0.348   |   0.277  |        0.247        |          0.137           |

### ONNX inference speeds

Exported to ONNX file format using `half=True` and testing using VisDrone Test dataset (1610 images). Single run for each model using Ryzen 5 1600 CPU.

```py
from ultralytics import YOLO()
model = YOLO('path/to/best.pt')
model.export(format='onnx',half=True)
...
model = YOLO('path/to/best.onnx')
model.val(data='VisDrone.yaml',batch=1) # results using this
```

| Model structure | preprocess | inference | postprocess |
| :----------------- | :-----------: |:----------: | :------------: | 
| YOLOv8n-ghost |     0.7 ms    |   67.2 ms  |     9.9 ms      |
| YOLOv8n            |     0.8 ms    |   85.7 ms  |     6.8 ms      |

### Weights file sizes

| Model structure  | PyTorch  | ONNX (half) |
| :------------------ | :--------: |:--------------: |
| YOLOv8n-ghost   |  3.7 MB  |    7.1  MB       |
| YOLOv8n-scratch |  6.2 MB  |   12.2 MB       |

### Summary
🆕👻🕵️‍♂️

Add a new YAML file `ultralytics/cfg/models/v8/yolov8-ghost.yaml` that defines the YOLOv8-ghost model, a lighter and faster variant of the YOLOv8 object detection model. The new model uses GhostConv and C3Ghost modules to reduce the number of parameters and FLOPs.

> _`YOLOv8-ghost` file_
> _A lighter detection model_
> _Autumn leaves falling_

### Walkthrough
*  Add a new model option `yolov8-ghost` that uses GhostConv and C3Ghost modules to reduce parameters and FLOPs ([link](https://github.com/ultralytics/ultralytics/pull/6749/files?diff=unified&w=0#diff-a2698aaa8231498c4bbb0afb15c14f386044efc87732c8d38108b9a5391a893cR1-R46))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introducing Ghost Convolutional Networks to YOLOv8 object detection models.

### 📊 Key Changes
- 🔨 Added a new `yolov8-ghost.yaml` configuration file.
- 📈 Defined scaling parameters for different sized models (YOLOv8n-ghost, YOLOv8s-ghost, YOLOv8m-ghost, YOLOv8l-ghost, YOLOv8x-ghost) indicating variations in depth, width, and max channels.
- 🏗️ Outlined the architecture of YOLOv8 with Ghost modules including Conv, GhostConv, C3Ghost, SPPF, and nn.Upsample layers.
- 🧠 Introduced an updated head for the model to include new Concat and C3Ghost layers for object detection across different scales (P3, P4, P5).

### 🎯 Purpose & Impact
- 🏎️ The integration of Ghost convolutions aims to improve the efficiency of YOLOv8 models, potentially reducing the computational cost while maintaining accuracy.
- 👁️ Users can now choose from different scales of the YOLOv8-ghost model depending on the demands of their task, balancing the trade-off between speed and accuracy.
- 💾 The new configuration file simplifies the creation and training of these leaner models for developers and hobbyists looking for high-performance object detection with reduced resource usage.

### CODE REVIEW

**Comment by ExtReMLapin** [comment]
P2 version

```yaml
# Ultralytics YOLO 🚀, AGPL-3.0 license
# YOLOv8 object detection model with P2-P5 outputs. For Usage examples see https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 80  # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024]
  s: [0.33, 0.50, 1024]
  m: [0.67, 0.75, 768]
  l: [1.00, 1.00, 512]
  x: [1.00, 1.25, 512]

# YOLOv8.0 backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]]  # 0-P1/2
  - [-1, 1, GhostConv, [128, 3, 2]]  # 1-P2/4
  - [-1, 3, C3Ghost, [128, True]]
  - [-1, 1, GhostConv, [256, 3, 2]]  # 3-P3/8
  - [-1, 6, C3Ghost, [256, True]]
  - [-1, 1, GhostConv, [512, 3, 2]]  # 5-P4/16
  - [-1, 6, C3Ghost, [512, True]]
  - [-1, 1, GhostConv, [1024, 3, 2]]  # 7-P5/32
  - [-1, 3, C3Ghost, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]]  # 9

# YOLOv8.0-p2 head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C3Ghost, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C3Ghost, [256]]  # 15 (P3/8-small)

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 2], 1, Concat, [1]]  # cat backbone P2
  - [-1, 3, C3Ghost, [128]]  # 18 (P2/4-xsmall)

  - [-1, 1, GhostConv, [128, 3, 2]]
  - [[-1, 15], 1, Concat, [1]]  # cat head P3
  - [-1, 3, C3Ghost, [256]]  # 21 (P3/8-small)

  - [-1, 1, GhostConv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C3Ghost, [512]]  # 24 (P4/16-medium)

  - [-1, 1, GhostConv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C3Ghost, [1024]]  # 27 (P5/32-large)

  - [[18, 21, 24, 27], 1, Detect, [nc]]  # Detect(P2, P3, P4, P5)
```

**Comment by fcakyon** [comment]
Amazing work @Burhan-Q 🚀
